### PR TITLE
Macro VIR printer (with compact mode)

### PR DESCRIPTION
**Pull request by utaal**
This removes the hand-written VIR printer and replaces it with a (mainly) macro-generated one.

At the same time, it supports `compact` printing mode from PR #203, but doesn't yet include `pretty_format`, which we can add back in later.

I prefer this solution for better maintainability, so changes to VIR don't always require a manual change to the printer.

### CODE REVIEW

File: source/vir/src/printer.rs [hunk 1]
```diff
@@ -1,24 +1,13 @@
 use crate::ast::*;
 use air::ast::Span;
```
Comment (channy412) [inline source/vir/src/printer.rs:2]:
```suggestion

```
I think it gives warning (unused import)

File: source/vir/src/printer.rs [hunk 2]
```diff
@@ -1,24 +1,13 @@
...
 use air::printer::macro_push_node;
-use air::printer::{clean_up_lines, str_to_node};
-use air::{node, nodes, nodes_vec};
+use air::printer::str_to_node;
```
Comment (channy412) [inline source/vir/src/printer.rs:4]:
```suggestion

```
I think this also gives the same warning

File: source/vir/src/ast.rs [hunk 3]
```diff
@@ -499,7 +501,7 @@ pub enum StmtX {
 /// Function parameter
 pub type Param = Arc<Spanned<ParamX>>;
 pub type Params = Arc<Vec<Param>>;
-#[derive(Debug, Clone)]
+#[derive(Debug, ToNode, Clone)]
```
Comment (channy412) [inline source/vir/src/ast.rs:504]: Do we also want to apply compact printing to function arguments and the return type?

File: source/vir/src/ast.rs [hunk 4]
```diff
@@ -406,6 +407,7 @@ pub enum ComputeMode {
 pub type Expr = Arc<SpannedTyped<ExprX>>;
 pub type Exprs = Arc<Vec<Expr>>;
 #[derive(Debug)]
+#[to_node_impl(name = ">")]
```
Comment (channy412) [inline source/vir/src/ast.rs:410]: Just to confrim, is ">" replacing "ExprX" in the VIR log?

**Review by channy412 — APPROVED** [review]
Thank you for working on the VIR printer!   This looks great to me, and looks much easier to maintain. I simply added a few comments. Thanks!

File: source/vir/src/ast.rs (hunk 3 above)
Comment (utaal) [reply source/vir/src/ast.rs:504]: Can you elaborate? The `ToNodeOpts` are propagated in all recursive calls, but maybe that's not what you mean.

File: source/vir/src/ast.rs (hunk 4 above)
Comment (utaal) [reply source/vir/src/ast.rs:410]: Yes!

File: source/vir/src/ast.rs (hunk 3 above)
Comment (channy412) [reply source/vir/src/ast.rs:504]: Using `compact` mode, the following code snippet generates the VIR log below. It seems like `ParamX`'s type information is logged, and I wonder if we want to remove these type information of function arguments and the return type.  If we choose to do that, I am guessing that we might add a handwritten `impl ToNode for ParamX`?

```
spec fn mul(a: u64, b: u64)  -> u64 {
    builtin::mul(a, b)
}
```

```
;; rust_verify/example/assorted_demo.rs:24:6: 24:36 (#23375)
(Function
 (Fun :path "mul." :trait_path None) :mode Spec :typ_bounds () :params (
  (Param :name "a~2" :typ (Typ Boxed (Typ Int (IntRange U 64))) :mode Spec :is_mut false)
  (Param :name "b~4" :typ (Typ Boxed (Typ Int (IntRange U 64))) :mode Spec :is_mut false)
 ) :ret (Param :name "%return" :typ (Typ Int (IntRange U 64)) :mode Spec :is_mut false) :require () :ensure () :body (> Block
  () (> Binary (BinaryOp Arith (ArithOp Mul) 24) (> Var "a~2") (> Var "b~4"))))
```

**Review by Chris-Hawblitzel — APPROVED** [review]
Sounds like a good idea.  My only comment is that the name `ToNode` isn't very self-explanatory when used in the `derive`s in `ast.rs`.  Can we name it something that makes clear that it's for debugging or printing, like `DebugNode` or `DebugPrint` or something?

File: source/vir/src/ast.rs (hunk 3 above)
Comment (utaal) [reply source/vir/src/ast.rs:504]: Thanks! I'd argue that type signatures for functions are still fairly critical info, so I'd print it in compact mode too (where I'd interpret `no_type` as "no types of expressions").

**Pull request merged**